sycl/source/detail/scheduler/commands.cpp

```diff
@@ -161,6 +226,190 @@ Command::Command(CommandType Type, QueueImplPtr Queue)
+                            nullptr);
+    }
+    // General comment - should we print error message for odd/bad
+    // events or ignore them for now?
+  }
+#endif
+}
+
+void Command::emitEdgeEventForEventDependence(Command *Cmd,
+                                              RT::PiEvent &PiEventAddr) {
+#ifdef XPTI_ENABLE_INSTRUMENTATION
+  // If we have failed to create an event to represent the Command, then we
+  // cannot emit an edge event. Bail early!
+  if (!MTraceEvent)
+    return;
+
+  if (xptiTraceEnabled()) {
+    if (Cmd && Cmd->MTraceEvent) {
+      // If the event is associated with a command, we use this command's trace
+      // event as the source of edge, hence modeling the control flow
+      emitEdgeEventForCommandDependence(Cmd, (void *)PiEventAddr, "Event",
+                                        false);
+
+    } else if (PiEventAddr) {
+      xpti::utils::StringHelper SH;
+      std::string AddressStr = SH.addressAsString(PiEventAddr);
+      // This is the case when it is a OCL event enqueued by the user or another
+      // event is registered by the runtime as a dependency The dependency on
+      // this occasion is an OCL event; so we build a virtual node in the graph
+      // with the event as the metadata for the node
+      std::string NodeName =
+          SH.nameWithAddressString("virtual_node", AddressStr);
+      // Node name is "virtual_node[<event_addr>]"
+      xpti::payload_t np(NodeName.c_str(), MAddress);
+      auto NodeEvent = xptiMakeEvent(NodeName.c_str(), &np,
+                                     (uint16_t)xpti_te::graph, xpti_at::active);
+      // Emit the virtual node first
+      xptiAddMetadata(NodeEvent, "kernel_name", NodeName.c_str());
+      xptiNotifySubscribers(MStreamID, (uint16_t)xpti_tp::node_create,
+                            cl::sycl::detail::g_sycl_graph_event, NodeEvent,
+                            nullptr);
+      // Create a new event for the edge
+      std::string EdgeName = SH.nameWithAddressString("Event", AddressStr);
+      xpti::payload_t p(EdgeName.c_str(), MAddress);
+      auto EdgeEvent = xptiMakeEvent(EdgeName.c_str(), &p,
+                                     (uint16_t)xpti_te::graph, xpti_at::active);
+      if (EdgeEvent && NodeEvent) {
+        // Source node represents the event and this event needs
+        // to be completed before target node can execute
+        xpti_td *TgtEvent = (xpti_td *)MTraceEvent;
+        EdgeEvent->source_id = NodeEvent->unique_id;
+        EdgeEvent->target_id = TgtEvent->unique_id;
+        xptiAddMetadata(EdgeEvent, "event", EdgeName.c_str());
+        xptiNotifySubscribers(MStreamID, (uint16_t)xpti_tp::edge_create,
+                              cl::sycl::detail::g_sycl_graph_event, EdgeEvent,
+                              nullptr);
+      }
+    }
+  }
+#endif
+}
+
+void Command::makeTraceEventProlog(void *MAddress) {
+#ifdef XPTI_ENABLE_INSTRUMENTATION
+  MTraceEventPrologComplete = true;
```

Comment:
Shouldn't we check that xpti is enabled?

Reply:
The method is always called from within an xpti enabled if block, but I can add another here so we are protected in case there is a use of this method by itself in the future.